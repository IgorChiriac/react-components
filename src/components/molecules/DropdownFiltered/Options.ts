import styled from 'styled-components';
import { colors } from '../../../constants';

export interface OptionsListProps {
  optionsListMaxHeight?: string;
}

const Options = styled.div<OptionsListProps>`
  z-index: 1;
  width: 100%;
  background: ${colors.white};
  border: 1px solid ${colors.brand};
  border-top: 0;
  position: absolute;
  overflow: auto;
  border-radius: 0 0 8px 8px;
  max-height: ${({ optionsListMaxHeight = 'initial' }) => optionsListMaxHeight};
  box-shadow: 0 0 4px 0 ${colors.brand};
`;

export default Options;
