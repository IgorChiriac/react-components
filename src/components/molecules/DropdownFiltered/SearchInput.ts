import styled, { css } from 'styled-components';
import { colors } from '../../../constants/colors';
import InputText from '../../atoms/InputText/InputText';
import { IInput } from '../../types';

export interface ISearchInputProps {
  isOpen?: boolean;
}

export const SearchInput = styled(InputText)<ISearchInputProps & IInput>`
  margin: 0;
  padding-left: 16px;
  padding-right: 32px;
  ${({ hasError }) => !hasError && 'margin-bottom: 0'};
  ${({ isOpen }) =>
    isOpen &&
    css`
      box-shadow: 0 0 4px 0 ${colors.brand};
      border-radius: 8px 8px 0 0;
      border-color: ${colors.brand};
    `};
`;

export const SearchInputWrap = styled.div`
  position: relative;
`;

export const SearchArrowWrap = styled.div`
  position: absolute;
  top: 50%;
  transform: translateY(-50%);
  right: 12px;
  cursor: pointer;
  display: flex;
`;
