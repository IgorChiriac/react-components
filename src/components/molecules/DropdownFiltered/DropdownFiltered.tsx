import Downshift, { ControllerStateAndHelpers, DownshiftProps } from 'downshift';
import React from 'react';
import styled from 'styled-components';
import { colors } from '../../../constants/colors';
import ErrorMessage from '../../atoms/ErrorMessage/ErrorMessage';
import InputLabel from '../../atoms/InputLabel/InputLabel';
import SizedContainer from '../../layout/SizedContainer/SizedContainer';
import Icon from '../../atoms/Icon/Icon';
import { faChevronDown } from '@fortawesome/free-solid-svg-icons';
import { IField, IInput } from '../../types';
import Option from './Option';
import Options, { IOptionsListProps } from './Options';
import { ISearchInputProps, SearchInput, SearchArrowWrap, SearchInputWrap } from './SearchInput';

export interface IDropdownItem {
  value: string;
}

export interface IDropdownFilteredProps
  extends DownshiftProps<IDropdownItem>,
    ISearchInputProps,
    IOptionsListProps,
    Omit<IInput, 'onSelect' | 'onChange' | 'children'>,
    IField {
  /**
   * Native props for the native label element.
   */
  labelProps?: React.LabelHTMLAttributes<HTMLLabelElement>;
  /**
   * Items for the options
   */
  items: IDropdownItem[];
}

const FieldError = styled(ErrorMessage)`
  margin-top: 5px;
`;

export default class DropdownFiltered extends React.PureComponent<IDropdownFilteredProps> {
  public render() {
    return (
      <SizedContainer size={this.props.inputSize}>
        <Downshift itemToString={item => (item ? item.value : '')} {...this.props}>
          {this.dropdownContent}
        </Downshift>
      </SizedContainer>
    );
  }

  private searchMatch = (itemValue: string, inputValue: string) =>
    itemValue && itemValue.toLowerCase().includes(inputValue.toLowerCase());

  private dropdownContent = (options: ControllerStateAndHelpers<IDropdownItem>) => {
    const {
      clearSelection,
      getInputProps,
      getItemProps,
      getLabelProps,
      getMenuProps,
      highlightedIndex,
      inputValue,
      isOpen,
      openMenu,
      closeMenu,
      selectedItem,
    } = options;

    const {
      errorMessage,
      items = [],
      label,
      optionsListMaxHeight,
      name,
      labelProps,
      isValid,
      disabled,
      onChange,
      onFocus,
      ...rest
    } = this.props;
    if (!name) {
      throw Error('Name must be set in inputProps. Check the docs.');
    }

    const filteredResults = items.filter(({ value }) => this.searchMatch(value, inputValue || ''));

    const showError = errorMessage && !isOpen;

    return (
      <div>
        {label && (
          <InputLabel {...getLabelProps({ ...labelProps })} htmlFor={`text-id-${name}`}>
            {label}
          </InputLabel>
        )}
        <SearchInputWrap>
          <SearchInput
            {...(getInputProps as any)({
              // Types doesn't allow custom data-* attrs
              onChange: (e: React.ChangeEvent<HTMLInputElement>) => {
                if (e.target.value === '') {
                  clearSelection();
                }
              },
              onFocus: () => {
                openMenu();
              },
              ...rest,
            })}
            name={name}
            id={`text-id-${name}`}
            isValid={isValid}
            hasError={showError}
            isOpen={isOpen}
            disabled={disabled}
          />
          <SearchArrowWrap>
            <Icon
              variant={faChevronDown}
              rotation={isOpen ? 180 : undefined}
              color={disabled ? colors.greyLight : colors.actionPlain}
              onClick={() => {
                if (!disabled) {
                  if (isOpen) {
                    closeMenu();
                  } else {
                    clearSelection(() => {
                      openMenu();
                    });
                  }
                }
              }}
              aria-label={isOpen ? 'close.menu' : 'open.menu'}
            />
          </SearchArrowWrap>
          {isOpen && (
            <Options {...getMenuProps()} optionsListMaxHeight={optionsListMaxHeight}>
              {filteredResults.map((item, index) => (
                <Option
                  {...getItemProps({ index, item })}
                  key={item.value}
                  highLighted={highlightedIndex === index}
                  selected={selectedItem === item}
                  disabled={disabled}
                >
                  {item.value}
                </Option>
              ))}
            </Options>
          )}
        </SearchInputWrap>
        {showError && <FieldError data-automation={`ZA.error-${name}`}>{errorMessage}</FieldError>}
      </div>
    );
  };
}
