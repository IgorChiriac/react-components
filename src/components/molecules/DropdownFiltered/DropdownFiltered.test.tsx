import React from 'react';
import { render } from 'react-testing-library';
import DropdownFiltered from './DropdownFiltered';

const items = [{ value: 'apple' }, { value: 'pear' }, { value: 'orange' }, { value: 'grape' }, { value: 'banana' }];

describe('<DropdownFiltered />', () => {
  it('renders the Dropdown with options with no a11y violations', async () => {
    const { container } = render(<DropdownFiltered items={items} label="Label" inputProps={{ name: 'test' }} />);
    expect(container.firstChild).toMatchSnapshot();
    // const results = await axe(container.innerHTML);
    // expect(results).toHaveNoViolations();
  });

  it('renders the DropdownFiltered with options and a default selected value', () => {
    const { container } = render(
      <DropdownFiltered defaultSelectedItem={{ value: 'orange' }} items={items} inputProps={{ name: 'test' }} />,
    );
    expect(container.firstChild).toMatchSnapshot();
  });
});
