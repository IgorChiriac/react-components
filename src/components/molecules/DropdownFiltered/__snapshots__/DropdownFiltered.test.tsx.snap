// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DropdownFiltered /> renders the Dropdown with options with no a11y violations 1`] = `
.c1 {
  display: block;
  margin: 0 0 10px;
  -webkit-letter-spacing: 0;
  -moz-letter-spacing: 0;
  -ms-letter-spacing: 0;
  letter-spacing: 0;
  color: #2C3236;
  font-family: "Open Sans",Roboto,Helvetica,Arial,sans-serif;
  font-weight: 700;
  font-size: 15px;
}

.c0 {
  max-width: 100%;
  width: 100%;
}

.c3 {
  width: 100%;
  -webkit-appearance: none;
  outline: none;
  border-radius: 8px;
  padding: 0 16px;
  height: 50px;
  font-size: 15px;
  font-weight: 400;
  color: #2C3236;
  border: 1px solid #818F9B;
  box-shadow: 0 0 4px 0 transparent;
  -webkit-transition-property: border,box-shadow;
  transition-property: border,box-shadow;
  -webkit-transition: 0.2s ease-in-out;
  transition: 0.2s ease-in-out;
}

.c3:focus {
  border: 1px solid #00D9C5;
  box-shadow: 0 0 4px 0 #00D9C5;
}

.c3::-webkit-input-placeholder {
  color: #D4D7D9;
  opacity: 1;
}

.c3::-moz-placeholder {
  color: #D4D7D9;
  opacity: 1;
}

.c3:-ms-input-placeholder {
  color: #D4D7D9;
  opacity: 1;
}

.c3::placeholder {
  color: #D4D7D9;
  opacity: 1;
}

.c3:disabled {
  background-color: #F7F7F7;
  color: #818F9B;
}

.c4 {
  margin: 0;
  padding-left: 16px;
  padding-right: 32px;
  margin-bottom: 0;
}

.c2 {
  position: relative;
}

.c5 {
  position: absolute;
  top: 50%;
  -webkit-transform: translateY(-50%);
  -ms-transform: translateY(-50%);
  transform: translateY(-50%);
  right: 12px;
  cursor: pointer;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

<div
  class="c0"
>
  <div
    aria-expanded="false"
    aria-haspopup="listbox"
    aria-labelledby="downshift-0-label"
    role="combobox"
  >
    <label
      class="c1"
      for="text-id-test"
      id="downshift-0-label"
    >
      Label
    </label>
    <div
      class="c2"
    >
      <input
        aria-autocomplete="list"
        aria-labelledby="downshift-0-label"
        autocomplete="off"
        class="c3 c4"
        id="text-id-test"
        name="test"
        value=""
      />
      <div
        class="c5"
      >
        <svg
          aria-hidden="true"
          aria-label="open.menu"
          class="svg-inline--fa fa-chevron-down fa-w-14 "
          color="#818F9B"
          data-icon="chevron-down"
          data-prefix="fas"
          focusable="false"
          role="img"
          viewBox="0 0 448 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M207.029 381.476L12.686 187.132c-9.373-9.373-9.373-24.569 0-33.941l22.667-22.667c9.357-9.357 24.522-9.375 33.901-.04L224 284.505l154.745-154.021c9.379-9.335 24.544-9.317 33.901.04l22.667 22.667c9.373 9.373 9.373 24.569 0 33.941L240.971 381.476c-9.373 9.372-24.569 9.372-33.942 0z"
            fill="currentColor"
          />
        </svg>
      </div>
    </div>
  </div>
</div>
`;

exports[`<DropdownFiltered /> renders the DropdownFiltered with options and a default selected value 1`] = `
.c0 {
  max-width: 100%;
  width: 100%;
}

.c2 {
  width: 100%;
  -webkit-appearance: none;
  outline: none;
  border-radius: 8px;
  padding: 0 16px;
  height: 50px;
  font-size: 15px;
  font-weight: 400;
  color: #2C3236;
  border: 1px solid #818F9B;
  box-shadow: 0 0 4px 0 transparent;
  -webkit-transition-property: border,box-shadow;
  transition-property: border,box-shadow;
  -webkit-transition: 0.2s ease-in-out;
  transition: 0.2s ease-in-out;
}

.c2:focus {
  border: 1px solid #00D9C5;
  box-shadow: 0 0 4px 0 #00D9C5;
}

.c2::-webkit-input-placeholder {
  color: #D4D7D9;
  opacity: 1;
}

.c2::-moz-placeholder {
  color: #D4D7D9;
  opacity: 1;
}

.c2:-ms-input-placeholder {
  color: #D4D7D9;
  opacity: 1;
}

.c2::placeholder {
  color: #D4D7D9;
  opacity: 1;
}

.c2:disabled {
  background-color: #F7F7F7;
  color: #818F9B;
}

.c3 {
  margin: 0;
  padding-left: 16px;
  padding-right: 32px;
  margin-bottom: 0;
}

.c1 {
  position: relative;
}

.c4 {
  position: absolute;
  top: 50%;
  -webkit-transform: translateY(-50%);
  -ms-transform: translateY(-50%);
  transform: translateY(-50%);
  right: 12px;
  cursor: pointer;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

<div
  class="c0"
>
  <div
    aria-expanded="false"
    aria-haspopup="listbox"
    aria-labelledby="downshift-1-label"
    role="combobox"
  >
    <div
      class="c1"
    >
      <input
        aria-autocomplete="list"
        aria-labelledby="downshift-1-label"
        autocomplete="off"
        class="c2 c3"
        id="text-id-test"
        name="test"
        value="orange"
      />
      <div
        class="c4"
      >
        <svg
          aria-hidden="true"
          aria-label="open.menu"
          class="svg-inline--fa fa-chevron-down fa-w-14 "
          color="#818F9B"
          data-icon="chevron-down"
          data-prefix="fas"
          focusable="false"
          role="img"
          viewBox="0 0 448 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M207.029 381.476L12.686 187.132c-9.373-9.373-9.373-24.569 0-33.941l22.667-22.667c9.357-9.357 24.522-9.375 33.901-.04L224 284.505l154.745-154.021c9.379-9.335 24.544-9.317 33.901.04l22.667 22.667c9.373 9.373 9.373 24.569 0 33.941L240.971 381.476c-9.373 9.372-24.569 9.372-33.942 0z"
            fill="currentColor"
          />
        </svg>
      </div>
    </div>
  </div>
</div>
`;
